### PR TITLE
Refactor code base

### PR DESCRIPTION
**Pull request by kevinhwang91**

### CODE REVIEW

**Comment by kevinhwang91** [comment]
I think `jump_text` option should be deleted and use the fixed value.

File: lua/neogen/generator.lua
Comment (danymat) [inline lua/neogen/generator.lua:2]: This should be moved to init.lua, as we want to warn users as soon as they install Neogen. Don't you think ?
Futhermore, I think the assert here is not good. We should use `notify`, with `ERROR`.

File: lua/neogen/generator.lua
Comment (danymat) [inline lua/neogen/generator.lua:90]: I don't think this line is required anymore. But we could replace with a ERROR notify

File: lua/neogen/generator.lua
Comment (danymat) [inline lua/neogen/generator.lua:176]: Or maybe we could move the previous comment here

File: lua/neogen/init.lua
Comment (danymat) [inline lua/neogen/init.lua:63]: I guess this is deleted because of the `__call` ?

File: lua/neogen/locators/default.lua
Comment (danymat) [inline lua/neogen/locators/default.lua:9]:
```suggestion
--- The default locator tries to find one of the nodes to match in the current node
--- If it does not find one, will fetch the parents until he finds one
---@param node_info Neogen.node_info a node informations
---@param nodes_to_match table a list of parent nodes to match
---@return userdata node one of the nodes to match directly above the given node
```

**Review by danymat — CHANGES_REQUESTED** [review]
Some other considerations: 

- Jumping after the last cursor position does not go back to position anterior to annotation generation and does not delete the placed marks.
- Jumping before the first cursor position should not work.

**Comment by danymat** [comment]
> I think `jump_text` option should be deleted and use the fixed value.

You're right, I let you remove all reference of jump_text (even in the comments)

**Comment by kevinhwang91** [comment]
> umping after the last cursor position does not go back to position anterior to annotation generation and does not delete the placed marks.
Jumping before the first cursor position should not work.

Have you tested the code? 
https://github.com/danymat/neogen/pull/64/files#diff-99fd81b9122c5191ab1e651ea4edf4a55e0442689881ef04bea2c2e4f76d7e4bR49-R57

File: lua/neogen/generator.lua
Comment (kevinhwang91) [reply lua/neogen/generator.lua:2]: require at init.lua can't delay load. I don't think users like the warning msg even though they don't use neogen right now.

File: lua/neogen/init.lua
Comment (kevinhwang91) [reply lua/neogen/init.lua:63]: As a plugin, don't use `_G`, unless we have to.

File: lua/neogen/locators/default.lua
Comment (kevinhwang91) [reply lua/neogen/locators/default.lua:9]: Yes, you are right.

**Comment by danymat** [comment]
In PR:
https://user-images.githubusercontent.com/5306901/152676651-4d595624-225f-4139-a9bc-b90ea4a8d1ef.mov

In main:

https://user-images.githubusercontent.com/5306901/152676692-dd2c873b-ebb4-4c97-b937-7275a12f8123.mov

By the way, I just found out that the marks do not delete themselves after we go outside of the annotation. Re-entering the inserted annotation should not allow jumping again:

https://user-images.githubusercontent.com/5306901/152676736-0163d63b-655d-4bcc-b0ae-1ca135ea6cac.mov

**Comment by kevinhwang91** [comment]
What's your setting for jumpping?

```vim
inoremap <C-j> <Cmd>lua require('neogen').jump_next()<CR>
inoremap <C-k> <Cmd>lua require('neogen').jump_prev()<CR>
```

**Comment by danymat** [comment]
I use cmp, like the one in readme.

**Comment by kevinhwang91** [comment]
update your config, replace `neogen.jumpable()` with `require('neogen').jumpable()`, `_G.neogen` is nil now.

**Comment by danymat** [comment]
> update your config, replace neogen.jumpable() with require('neogen').jumpable(), _G.neogen is nil now.

Right, I will try it, then you would need to update the readme too for these changes

> TBH, I think your setting in README is weird.

This is how LuaSnip, nvim-cmp use concurrent keys. 
See: https://github.com/hrsh7th/nvim-cmp/wiki/Example-mappings, https://github.com/L3MON4D3/LuaSnip#keymaps 
why this is the best idea for concurrent keys.

**Comment by kevinhwang91** [comment]
I have known the issue about ending jumping, I will push the commit later.

**Review by danymat — APPROVED** [review]

File: lua/neogen/generator.lua
Comment (danymat) [reply lua/neogen/generator.lua:2]: Right, okay to keep it there. What about using notify then instead of `assert` ?

**Review by danymat — APPROVED** [review]

**Comment by danymat** [comment]
Still unresolved :
- https://github.com/danymat/neogen/pull/64#issuecomment-1030801350
- https://github.com/danymat/neogen/pull/64#discussion_r800170043

File: lua/neogen/generator.lua
Comment (kevinhwang91) [reply lua/neogen/generator.lua:2]: Yes, it is better than throwing the error make users feel panic.

**Review by danymat — COMMENTED** [review]
Nice, protip for future: no need to update .txt help file as it's auto generated from comments

**Comment by kevinhwang91** [comment]
> Nice, protip for future: no need to update .txt help file as it's auto generated from comments

Oh, I forget it, but I edit them using quickfix very fast.

**Comment by danymat** [comment]
Good morning, is this ready to merge ? If so, I will add some documentation before merging, and provide the commit as `BREAKING`

**Comment by kevinhwang91** [comment]
Yes

File: lua/neogen/init.lua
Comment (danymat) [inline lua/neogen/init.lua:133]: @kevinhwang91 Lua lsp is throwing me a warning about the function that accepts 0 variables. Maybe it's because of the `__call`. Is it possible to stop the warning maybe by making the return of the require a function that accepts 2 variables, or something else ?

**Comment by danymat** [comment]
@kevinhwang91 It seems that I still have the ability to jump to the annotations after I exit the annotations and go back again. Can you reproduce ?

**Comment by kevinhwang91** [comment]
> @kevinhwang91 It seems that I still have the ability to jump to the annotations after I exit the annotations and go back again. Can you reproduce ?
> 
> [#64 (comment)](https://github.com/danymat/neogen/pull/64#issuecomment-1030803813) (last video)

Have added range to check jumpable.

**Comment by kevinhwang91** [comment]
The issue should be caused by your config. Neither coc.nvim nor nvim-cmp can't reproduce it for me.

**Comment by danymat** [comment]
Right, it seems to be resolved somehow, weird

**Comment by danymat** [comment]
Merged, thanks @kevinhwang91 ❤️

**Pull request merged**
